[RISCV][DAGCombiner] Fix potential missed combine in VL->VW extension

OMG-link opened the pull request:
The previous implementation of `combineOp_VLToVWOp_VL` manually replaced old
nodes with newly created widened nodes, but only added the new node itself to
the `DAGCombiner` worklist. Since the users of the new node were not added,
some combine opportunities could be missed when external `DAGCombiner` passes
expected those users to be reconsidered.

This patch replaces the custom replacement logic with a call to
`DCI.CombineTo()`, which performs node replacement in a way consistent with
`DAGCombiner::Run`:
- Replace all uses of the old node.
- Add the new node and its users to the worklist.
- Clean up unused nodes when appropriate.

Using `CombineTo` ensures that `combineOp_VLToVWOp_VL` behaves consistently with
the standard `DAGCombiner` update model, avoiding discrepancies between the
private worklist inside this routine and the global worklist managed by the
combiner.

This resolves missed combine cases involving VL -> VW operator widening.

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/vscale-vw-web-simplification.ll:
These tests should be kept, just with the default RUN lines without the -riscv-lower-ext-max-web-size flag. We'll miss the test coverage otherwise

topperc commented on llvm/test/CodeGen/RISCV/rvv/vscale-vw-web-simplification.ll:
```diff
@@ -595,11 +383,12 @@ define <vscale x 4 x i32> @mismatched_extend_sub_add_commuted(<vscale x 4 x i16>
-; FOLDING-NEXT:    vwsub.wv v12, v10, v9
-; FOLDING-NEXT:    vwadd.wv v10, v10, v9
+; FOLDING-NEXT:    vwsub.wv v10, v10, v9
+; FOLDING-NEXT:    vwaddu.wv v12, v12, v8
```
Yes this a regression. This undoes an improvement I made last week.

OMG-link replied:
The worklist inside `combineOp_VLToVWOp_VL` and the one in `DAGCombiner` are supposed to serve the same purpose. I need some time to understand why removing the worklist in `combineOp_VLToVWOp_VL` would cause your change to stop working. This should not happen.

topperc commented:
This patch fixes the reported issue and still allows the whole web to be handled at once. The important part is that `combineTo` also adds the users of the replaced node to the worklist.

```
diff --git a/llvm/lib/Target/RISCV/RISCVISelLowering.cpp b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
index 38cce26e44af..8ba1215561dc 100644
--- a/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
+++ b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
@@ -18300,8 +18300,7 @@ static SDValue combineOp_VLToVWOp_VL(SDNode *N,
     }
   }
   for (std::pair<SDValue, SDValue> OldNewValues : ValuesToReplace) {
-    DAG.ReplaceAllUsesOfValueWith(OldNewValues.first, OldNewValues.second);
-    DCI.AddToWorklist(OldNewValues.second.getNode());
+    DCI.CombineTo(OldNewValues.first.getNode(), OldNewValues.second);
   }
   return InputRootReplacement;
 }
```

OMG-link committed:
[RVV] Add test for missed VWMACC combine

Add a minimal reproducer for consecutive vwmacc-like operations
to illustrate that the previous DAG combine logic may miss combining
mul+add chains into a single vwmacc.vx instruction.

topperc commented:
> I looked at the changes @topperc made last week. To stay compatible with those changes, the only approach I can currently think of is as mentioned: keep the internal worklist, but also add the users of the updated nodes to the outer worklist.
> 
> In any case, I don’t think maintaining another worklist internally is good practice. But for now, it seems to be the only option we have.

@preames and I had a conversation about this last week when I was preparing my change. We think the right long term direction is to move this entire transform to a MachineIR pass. That would still need some kind of worklist and visited set, but it wouldn't be nested inside DAG combine. It would also allow the optimization to be shared with GlobalISel.

topperc commented:
@OMG-link Do you mind if I take your test and create a PR with my change?

OMG-link commented:
> @OMG-link Do you mind if I take your test and create a PR with my change?

I was planning to change my PR tomorrow because I can't connect to my server right now. But of course, you can do it if you want.

OMG-link committed:
[RISCV][DAGCombiner] Fix missed combines in combineOp_VLToVWOp_VL

The previous implementation of combineOp_VLToVWOp_VL manually replaced old
nodes with newly created widened nodes, but only added the new node itself to
the DAGCombiner worklist. Since the users of the new node were not added,
some combine opportunities could be missed when external DAGCombiner passes
expected those users to be reconsidered.

This patch replaces the custom replacement logic with a call to
DCI.CombineTo(), which performs node replacement in a way consistent with
DAGCombiner::Run:
- Replace all uses of the old node.
- Add the new node and its users to the worklist.
- Clean up unused nodes when appropriate.

Using CombineTo ensures that combineOp_VLToVWOp_VL behaves consistently with
the standard DAGCombiner update model, avoiding discrepancies between the
private worklist inside this routine and the global worklist managed by the
combiner.

This resolves missed combine cases involving VL -> VW operator widening.

OMG-link commented:
I have rebuilt the commit. This version keeps the private worklist and, when inserting newly created nodes into the outer worklist, preserves the same insertion logic used by the outer worklist. Although this leads to duplicated code, it’s the only way to maintain consistent worklist behavior as long as the private worklist exists.

In addition, this version of the changes does not affect the results of any existing tests; it only changes the new test case from *mul+add* to *macc*.

```diff
diff --git a/llvm/lib/Target/RISCV/RISCVISelLowering.cpp b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
index 38cce26e44af..75fab219b6c6 100644
--- a/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
+++ b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
@@ -18300,8 +18300,24 @@ static SDValue combineOp_VLToVWOp_VL(SDNode *N,
     }
   }
   for (std::pair<SDValue, SDValue> OldNewValues : ValuesToReplace) {
-    DAG.ReplaceAllUsesOfValueWith(OldNewValues.first, OldNewValues.second);
-    DCI.AddToWorklist(OldNewValues.second.getNode());
+    // This helper function does what is done for RV in `DAGCombiner::Run`.
+    [&DAG, &DCI](SDNode *N, SDValue RV) -> void {
+      if (N->getNumValues() == RV->getNumValues())
+        DAG.ReplaceAllUsesWith(N, RV.getNode());
+      else {
+        assert(N->getValueType(0) == RV.getValueType() &&
+               N->getNumValues() == 1 && "Type mismatch");
+        DAG.ReplaceAllUsesWith(N, &RV);
+      }
+
+      if (RV.getOpcode() != ISD::EntryToken) {
+        DCI.AddToWorklist(RV.getNode());
+        for (SDNode *Node : RV.getNode()->users())
+          DCI.AddToWorklist(Node);
+      }
+
+      DCI.recursivelyDeleteUnusedNodes(N);
+    }(OldNewValues.first.getNode(), OldNewValues.second);
   }
   return InputRootReplacement;
 }
```

However, I noticed that in the initial version of the changes, there was the following transformation:

```diff
diff --git a/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vwmul.ll b/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vwmul.ll
index 5ec5ce11f85a..94c3138fd330 100644
--- a/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vwmul.ll
+++ b/llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vwmul.ll
@@ -866,16 +866,16 @@ define <2 x i16> @vwmul_v2i16_multiuse(ptr %x, ptr %y, ptr %z, ptr %w) {
 ; CHECK-NEXT:    vle8.v v8, (a1)
 ; CHECK-NEXT:    vle8.v v9, (a2)
 ; CHECK-NEXT:    vsext.vf2 v10, v8
-; CHECK-NEXT:    vsext.vf2 v11, v9
-; CHECK-NEXT:    vdivu.vv v9, v10, v11
-; CHECK-NEXT:    vle8.v v10, (a0)
+; CHECK-NEXT:    vsext.vf2 v8, v9
+; CHECK-NEXT:    vdivu.vv v8, v10, v8
+; CHECK-NEXT:    vle8.v v9, (a0)
 ; CHECK-NEXT:    vle8.v v11, (a3)
-; CHECK-NEXT:    vsetvli zero, zero, e8, mf8, ta, ma
-; CHECK-NEXT:    vwmul.vv v12, v10, v11
-; CHECK-NEXT:    vwmul.vv v10, v8, v11
-; CHECK-NEXT:    vsetvli zero, zero, e16, mf4, ta, ma
-; CHECK-NEXT:    vor.vv v8, v12, v10
-; CHECK-NEXT:    vor.vv v8, v8, v9
+; CHECK-NEXT:    vsext.vf2 v12, v9
+; CHECK-NEXT:    vsext.vf2 v9, v11
+; CHECK-NEXT:    vmul.vv v11, v12, v9
+; CHECK-NEXT:    vmul.vv v9, v10, v9
+; CHECK-NEXT:    vor.vv v9, v11, v9
+; CHECK-NEXT:    vor.vv v8, v9, v8
 ; CHECK-NEXT:    ret
   %a = load <2 x i8>, ptr %x
   %b = load <2 x i8>, ptr %y
```

The new version can no longer discover this transformation (it introduces two additional `vsetvli` instructions, but this is clearly not the reason the current code fails to detect this opportunity). Perhaps we should open a separate PR to discuss this issue.

@topperc

OMG-link commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18300,8 +18300,24 @@ static SDValue combineOp_VLToVWOp_VL(SDNode *N,
   for (std::pair<SDValue, SDValue> OldNewValues : ValuesToReplace) {
-    DAG.ReplaceAllUsesOfValueWith(OldNewValues.first, OldNewValues.second);
-    DCI.AddToWorklist(OldNewValues.second.getNode());
+    // This helper function do as what is done for RV in `DAGCombiner::Run`.
```
What I originally intended was to keep the logic here completely consistent with how the outer worklist handles newly added nodes. However, considering that calling `DCI.CombineTo` seems to implement similar logic, the maintainer of the outer worklist will likely to keep these two pieces of logic in sync. I’ll go with your suggestion.

topperc reviewed (approved):
LGTM

topperc commented on llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll:
```diff
@@ -0,0 +1,54 @@
+  %add1 = add <32 x i16> %mul1, %acc1
+
+  store <32 x i16> %add0, <32 x i16>* %acc0_ptr, align 1
+  store <32 x i16> %add1, <32 x i16>* %acc1_ptr, align 1
```
Typed pointers are deprecated.

```suggestion
  store <32 x i16> %add0, ptr %acc0_ptr, align 1
  store <32 x i16> %add1, ptr %acc1_ptr, align 1
```

OMG-link replied:
Updated. @topperc

topperc commented on llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll:
```diff
@@ -0,0 +1,54 @@
+  %acc0 = load <32 x i16>, <32 x i16>* %acc0_ptr, align 1
+  %acc1 = load <32 x i16>, <32 x i16>* %acc1_ptr, align 1
+
+  %v8 = load <32 x i8>, <32 x i8>* %vptr, align 1
```
Typed pointers are deprecated

```suggestion
  %acc0 = load <32 x i16>, ptr %acc0_ptr, align 1
  %acc1 = load <32 x i16>, ptr %acc1_ptr, align 1

  %v8 = load <32 x i8>, ptr %vptr, align 1
```

topperc commented on llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll:
```diff
@@ -0,0 +1,54 @@
+  %v16 = sext <32 x i8> %v8 to <32 x i16>
+
+  %s0_ptr = getelementptr i8, i8* %scalars, i32 0
+  %s0_i8 = load i8, i8* %s0_ptr, align 1
```
```suggestion
  %s0_i8 = load i8, ptr %s0_ptr, align 1
```

topperc reviewed (approved):
LGTM

topperc commented on llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll:
```diff
@@ -0,0 +1,54 @@
+  %s1_ptr = getelementptr i8, ptr %scalars, i32 1
+  %s1_i8 = load i8, ptr %s1_ptr, align 1
+  %s1_i16 = sext i8 %s1_i8 to i16
+  %tmp1 = insertelement <32 x i16> undef, i16 %s1_i16, i32 0
```
```suggestion
  %tmp1 = insertelement <32 x i16> poison, i16 %s1_i16, i32 0
```

topperc commented on llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll:
```diff
@@ -0,0 +1,54 @@
+  %s1_i8 = load i8, ptr %s1_ptr, align 1
+  %s1_i16 = sext i8 %s1_i8 to i16
+  %tmp1 = insertelement <32 x i16> undef, i16 %s1_i16, i32 0
+  %splat1 = shufflevector <32 x i16> %tmp1, <32 x i16> undef, <32 x i32> zeroinitializer
```
```suggestion
  %splat1 = shufflevector <32 x i16> %tmp1, <32 x i16> poison, <32 x i32> zeroinitializer
```

OMG-link replied:
It looks like there were 4 uses of `undef` but only 3 of them were changed in your suggestion. I don't know how to append a new change and commit them together online, so I rebuilt them locally. Please take a look.

```diff
diff --git a/llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll b/llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll
index 8dd8e07dbb6e..943d8d2409ff 100644
--- a/llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll
+++ b/llvm/test/CodeGen/RISCV/rvv/combine-vl-vw-macc.ll
@@ -31,16 +31,16 @@ entry:
   %s0_ptr = getelementptr i8, ptr %scalars, i32 0
   %s0_i8 = load i8, ptr %s0_ptr, align 1
   %s0_i16 = sext i8 %s0_i8 to i16
-  %tmp0 = insertelement <32 x i16> undef, i16 %s0_i16, i32 0
-  %splat0 = shufflevector <32 x i16> %tmp0, <32 x i16> undef, <32 x i32> zeroinitializer
+  %tmp0 = insertelement <32 x i16> poison, i16 %s0_i16, i32 0
+  %splat0 = shufflevector <32 x i16> %tmp0, <32 x i16> poison, <32 x i32> zeroinitializer
   %mul0 = mul <32 x i16> %splat0, %v16
   %add0 = add <32 x i16> %mul0, %acc0
 
   %s1_ptr = getelementptr i8, ptr %scalars, i32 1
   %s1_i8 = load i8, ptr %s1_ptr, align 1
   %s1_i16 = sext i8 %s1_i8 to i16
-  %tmp1 = insertelement <32 x i16> undef, i16 %s1_i16, i32 0
-  %splat1 = shufflevector <32 x i16> %tmp1, <32 x i16> undef, <32 x i32> zeroinitializer
+  %tmp1 = insertelement <32 x i16> poison, i16 %s1_i16, i32 0
+  %splat1 = shufflevector <32 x i16> %tmp1, <32 x i16> poison, <32 x i32> zeroinitializer
   %mul1 = mul <32 x i16> %splat1, %v16
   %add1 = add <32 x i16> %mul1, %acc1
```

OMG-link commented:
> The PR description needs to be updated before landing, since it will be used for the commit message when merging

Updated.
